### PR TITLE
SONAR-8740 add organization to response of api favorites search

### PR DESCRIPTION
**Pull request by sns-seb**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/favorite/ws/SearchAction.java [hunk 1]
```diff
@@ -154,47 +138,35 @@ public SearchResults build() {
...
-      };
-    }
+  private void addFavorites(SearchResponse.Builder builder, SearchResults results) {
+    Favorite.Builder favoriteBuilder = Favorite.newBuilder();
+    results.favorites.stream()
+      .map(componentDto -> toWsFavorite(favoriteBuilder, componentDto))
+      .forEach(builder::addFavorites);
+  }
 
+  private Favorite toWsFavorite(Favorite.Builder builder, ComponentDto componentDto) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/favorite/ws/SearchAction.java]: MAJOR Make "toWsFavorite" a "static" method. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

File: server/sonar-server/src/main/java/org/sonar/server/favorite/ws/SearchAction.java [hunk 2]
```diff
@@ -154,47 +138,35 @@ public SearchResults build() {
...
+  private SearchResponse toSearchResponse(SearchResults searchResults) {
+    SearchResponse.Builder builder = SearchResponse.newBuilder();
+    addPaging(builder, searchResults);
+    addFavorites(builder, searchResults);
+    return builder.build();
+  }
 
-    private Consumer<SearchResults> addPaging() {
-      return results -> response.setPaging(Common.Paging.newBuilder()
+  private void addPaging(SearchResponse.Builder builder, SearchResults results) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/favorite/ws/SearchAction.java]: MAJOR Make "addPaging" a "static" method. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 2 major

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

File: server/sonar-server/src/main/java/org/sonar/server/favorite/ws/SearchAction.java [hunk 3]
```diff
@@ -75,126 +74,97 @@ public void define(WebService.NewController context) {
...
-      };
-    }
+  private void addFavorites(SearchResponse.Builder builder, SearchResults results) {
+    Favorite.Builder favoriteBuilder = Favorite.newBuilder();
+    results.favorites.stream()
+      .map(componentDto -> toWsFavorite(favoriteBuilder, results, componentDto))
+      .forEach(builder::addFavorites);
+  }
 
+  private Favorite toWsFavorite(Favorite.Builder builder, SearchResults results, ComponentDto componentDto) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/favorite/ws/SearchAction.java]: MAJOR Make "toWsFavorite" a "static" method. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

**Comment by simonbrandhof** [comment]
Minor - typo in message of commit ed54e5b ! ("lamdas")

File: server/sonar-server/src/main/java/org/sonar/server/favorite/ws/SearchAction.java [hunk 4]
```diff
@@ -72,11 +71,9 @@ public void define(WebService.NewController context) {
...
   @Override
   public void handle(Request request, Response response) throws Exception {
-    SearchResponse wsResponse = Stream.of(request)
-      .map(SearchAction::toWsRequest)
-      .map(this::toSearchResults)
-      .map(this::toSearchResponse)
-      .collect(Collectors.toOneElement());
+    SearchRequest searchRequest = toWsRequest(request);
+    SearchResults searchResults = toSearchResults(searchRequest);
+    SearchResponse wsResponse = toSearchResponse(searchResults);
```
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/favorite/ws/SearchAction.java]: Excellent. +1 !

File: server/sonar-server/src/main/java/org/sonar/server/favorite/ws/SearchAction.java [hunk 5]
```diff
@@ -86,83 +89,78 @@ private static SearchRequest toWsRequest(Request request) {
   private SearchResults toSearchResults(SearchRequest request) {
     userSession.checkLoggedIn();
     try (DbSession dbSession = dbClient.openSession(false)) {
-      SearchResults.Builder builder = SearchResults.newBuilder(request);
-      addAuthorizedProjectUuids(dbSession, builder);
-      addFavorites(builder);
-      return builder.build();
+      List<ComponentDto> authorizedFavorites = favoriteFinder.list().stream()
```
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/favorite/ws/SearchAction.java]: As `favoriteFinder#list()` is used only by this call, I wonder if it should not manage permission filtering.

File: server/sonar-server/src/main/java/org/sonar/server/favorite/ws/SearchAction.java [hunk 6]
```diff
@@ -86,83 +89,78 @@ private static SearchRequest toWsRequest(Request request) {
...
+        .filter(isAuthorized(dbSession))
+        .collect(Collectors.toList());
+      Paging paging = Paging.forPageIndex(request.getPage()).withPageSize(request.getPageSize()).andTotal(authorizedFavorites.size());
+      List<ComponentDto> displayedFavorites = authorizedFavorites.stream()
+        .skip(paging.offset())
+        .limit(paging.pageSize())
+        .collect(Collectors.toList());
```
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/favorite/ws/SearchAction.java]: you could optimize with `Collectors.toList(paging.pageSize())`.

File: server/sonar-server/src/main/java/org/sonar/server/favorite/ws/SearchAction.java [hunk 7]
```diff
@@ -86,83 +89,78 @@ private static SearchRequest toWsRequest(Request request) {
...
+      Map<String, OrganizationDto> organizationsByUuid = getOrganizationsOfComponents(dbSession, displayedFavorites);
+      return new SearchResults(paging, displayedFavorites, organizationsByUuid);
     }
   }
 
-  private void addFavorites(SearchResults.Builder builder) {
-    builder.allFavorites = favoriteFinder.list();
+  private Predicate<ComponentDto> isAuthorized(DbSession dbSession) {
+    Collection<String> rootProjectsUuids = dbClient.authorizationDao().selectAuthorizedRootProjectsUuids(dbSession, userSession.getUserId(), UserRole.USER);
```
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/favorite/ws/SearchAction.java]: The algorithm is to load all the projects visible by user (so potentially a huge number) and then to filter favorites ? If yes, then what about calling `userPermission#hasComponentPermission()` which contains a cache ?

**Pull request merged**